[WebKit] Add an option key to prevent network loads when serializing attributed strings

whsieh opened the pull request:
#### 39925c5d4c4ef3ff0ba785efc1b5d46a242ecf99
<pre>
[WebKit] Add an option key to prevent network loads when serializing attributed strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=264745">https://bugs.webkit.org/show_bug.cgi?id=264745</a>
<a href="https://rdar.apple.com/118309399">rdar://118309399</a>

Reviewed by Tim Horton.

Add support for a `NSAttributedStringDocumentReadingOptionKey` which allows clients to serialize
attributed strings from web content, without loading any requests over the network. This is
particularly ideal for clients such as UIFoundation, when serializing HTML markup and/or web archive
data, which otherwise has the potential to trigger arbitrary resource loads upon conversion.

Note that we add an opt-in flag for network loads, because we want to eventually default to *not*
loading content from over the network when serializing attributed strings. Given this eventual plan,
it makes more sense for an app to need to explicitly opt _in_ to network loads, rather than opt out.

Test: WKWebView.DoNotAllowNetworkLoads

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
(+[_WKAttributedStringWebViewCache configuration]):

Respect the new `_WKAllowNetworkLoadsOption` option by setting `_allowedNetworkHosts` to an empty
set on the configuration, if the value is false-y.

(+[_WKAttributedStringWebViewCache maybeUpdateShouldAllowNetworkLoads:]):
(+[_WKAttributedStringWebViewCache maybeConsumeBundlePaths:]):
(+[_WKAttributedStringWebViewCache invalidateGlobalConfigurationIfNeeded:]):

Refactor this logic to update the global configuration (and clear any cached web views if needed),
if either network load disablement or the allowed file URL list changes.

(+[NSAttributedString _loadFromHTMLWithOptions:contentLoader:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/NSAttributedStringPrivate.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadDelegate.mm:

Lift the `TestResourceLoadDelegate` class out of this API test and into a separate helper file, so
that we can use it in a different API test suite (see below).

(-[TestResourceLoadDelegate webView:resourceLoad:didSendRequest:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didPerformHTTPRedirection:newRequest:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveChallenge:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveResponse:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didCompleteWithError:response:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:

Add a new API test to exercise the case where `_WKAllowNetworkLoadsOption` is set to `@NO`, by
adding a resource load delegate to the web view to listen for any attempted network access. When
this new option is set, we should never get a call to the `didSendRequest:` delegate method.

* Tools/TestWebKitAPI/cocoa/TestResourceLoadDelegate.h: Added.
* Tools/TestWebKitAPI/cocoa/TestResourceLoadDelegate.mm: Added.
(-[TestResourceLoadDelegate webView:resourceLoad:didSendRequest:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didPerformHTTPRedirection:newRequest:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveChallenge:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveResponse:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didCompleteWithError:response:]):

Canonical link: <a href="https://commits.webkit.org/270753@main">https://commits.webkit.org/270753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c65d8fa04d06942d1944bb76216b6a0ac8b56632

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4879 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28366 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24045 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26601 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6662 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2293 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24052 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26526 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3726 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22593 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28944 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3329 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23559 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29622 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23943 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23957 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27512 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3357 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1566 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4780 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6323 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3835 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3682 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
LGTM

---

Editorial nit: when I was reading the commit log I assumed the opt-in flag you referred to was for `Add support for a 'NSAttributedStringDocumentReadingOptionKey' which allows clients to serialize
attributed strings from web content, without loading any requests over the network.`, but it's the other way around and the flag lets you opt-in to network loads. Maybe change the second paragraph to start like so?

```
Note that we add an opt-in flag for network loads, because we want to ...
```

whsieh committed:
[WebKit] Add an option key to prevent network loads when serializing attributed strings
https://bugs.webkit.org/show_bug.cgi?id=264745
rdar://118309399

Reviewed by Tim Horton.

Add support for a `NSAttributedStringDocumentReadingOptionKey` which allows clients to serialize
attributed strings from web content, without loading any requests over the network. This is
particularly ideal for clients such as UIFoundation, when serializing HTML markup and/or web archive
data, which otherwise has the potential to trigger arbitrary resource loads upon conversion.

Note that we add an opt-in flag for network loads, because we want to eventually default to *not*
loading content from over the network when serializing attributed strings. Given this eventual plan,
it makes more sense for an app to need to explicitly opt _in_ to network loads, rather than opt out.

Test: WKWebView.DoNotAllowNetworkLoads

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
(+[_WKAttributedStringWebViewCache configuration]):

Respect the new `_WKAllowNetworkLoadsOption` option by setting `_allowedNetworkHosts` to an empty
set on the configuration, if the value is false-y.

(+[_WKAttributedStringWebViewCache maybeUpdateShouldAllowNetworkLoads:]):
(+[_WKAttributedStringWebViewCache maybeConsumeBundlePaths:]):
(+[_WKAttributedStringWebViewCache invalidateGlobalConfigurationIfNeeded:]):

Refactor this logic to update the global configuration (and clear any cached web views if needed),
if either network load disablement or the allowed file URL list changes.

(+[NSAttributedString _loadFromHTMLWithOptions:contentLoader:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/NSAttributedStringPrivate.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadDelegate.mm:

Lift the `TestResourceLoadDelegate` class out of this API test and into a separate helper file, so
that we can use it in a different API test suite (see below).

(-[TestResourceLoadDelegate webView:resourceLoad:didSendRequest:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didPerformHTTPRedirection:newRequest:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveChallenge:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveResponse:]): Deleted.
(-[TestResourceLoadDelegate webView:resourceLoad:didCompleteWithError:response:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:

Add a new API test to exercise the case where `_WKAllowNetworkLoadsOption` is set to `@NO`, by
adding a resource load delegate to the web view to listen for any attempted network access. When
this new option is set, we should never get a call to the `didSendRequest:` delegate method.

* Tools/TestWebKitAPI/cocoa/TestResourceLoadDelegate.h: Added.
* Tools/TestWebKitAPI/cocoa/TestResourceLoadDelegate.mm: Added.
(-[TestResourceLoadDelegate webView:resourceLoad:didSendRequest:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didPerformHTTPRedirection:newRequest:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveChallenge:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didReceiveResponse:]):
(-[TestResourceLoadDelegate webView:resourceLoad:didCompleteWithError:response:]):

Canonical link: https://commits.webkit.org/270753@main
